### PR TITLE
Adding support for dynamic type

### LINKED ISSUE #59: Support for dynamic type
We (me and @fsaragoca) were thinking about having snapshot tests according to [content size category](https://developer.apple.com/reference/uikit/uiapplication/1623048-preferredcontentsizecategory?language=objc) as testing Dynamic Type manually is boring and no one seems to remember doing it when implementing a view/screen.

Initially I thought about making a separated component that used Nimble-Snapshots, but soon enough I've found myself needing `internal` methods, such as `_sanitizedTestName`.

I'd like to think what you folks think about this, if this would be a welcome addition, if it should be a subspec, etc.

The only downside for adding this to the library would be that I needed to swizzle (I actually used OCMock) `-[UIApplication preferredContentSizeCategory]` and `-[UIFont preferredFontForTextStyle]`. That's why I think a subspec would be better.

Also, the current implementation posts an `UIContentSizeCategoryDidChangeNotification` when the `preferredContentSizeCategory` is mocked, so the views/view controllers can update their content. This *might* cause trouble if whoever is writing the tests don't know about this implementation detail.

I also need names better than `recordDynamicTypeSnapshot` and `haveValidDynamicTypeSnapshot`. I thought about adding an extra parameter to the existing functions, but I think the default setting when someone wants to test Dynamic Type should be to test all the content size categories.

In short I think these are the options:

1. Add this feature to Nimble-Snapshots
2. Add this feature to Nimble-Snapshots in a subspec
3. Make this another component that uses Nimble-Snapshots, but change the visibility of needed methods
4. Make this another component that uses Nimble-Snapshots, but reimplement needed methods ¯\_(ツ)_/¯

### PR DESCRIPTION
**Pull request by marcelofabri**
This will fix #59 eventually

Missing:

- [x]  Docs
- [x]  Device Agnostic functions
- [x]  Pretty syntax with `==`
- [x] Check with Swift 3.0
- [x] Update CHANGELOG

Important points:

- I had to change the tests to run on both iOS 9 and 10, as the implementation is a bit different between them
- This probably don't work with Swift 3 (is there an easy way to check that?)
- I've bumped the Example project to iOS 9

Questions:

- Should we swizzle `UITraitCollection` on iOS 10 to return the mocked `preferredContentSizeCategory`?
- Should we post a notification when stop mocking to return to the old `preferredContentSizeCategory` value?

### CODE REVIEW

File: Nimble-Snapshots.podspec [hunk 1]
```diff
@@ -12,9 +12,22 @@ Pod::Spec.new do |s|
...
+  s.subspec "Core" do |ss|
+    ss.source_files  = "HaveValidSnapshot.swift", "PrettySyntax.swift", "NimbleSnapshotsConfiguration.swift"
+    ss.dependency "FBSnapshotTestCase", "~> 2.0"
+    ss.dependency "Nimble"
+    ss.dependency "Quick"
+  end
+
+  s.subspec "DynamicType" do |ss|
+    ss.source_files  = "DynamicType/*.{swift,m,h}"
+    # ss.private_header_files = "DynamicType/MockedApplication.h"
```
Comment (marcelofabri) [inline Nimble-Snapshots.podspec]: If I uncomment this, the Swift file can't access that class. Is there a way to make this visible to our library but hidden for outside consumers?

File: Bootstrap/BootstrapTests/DynamicTypeTests.swift [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        describe("in some context", {
+            var view: UIView!
+
+            beforeEach {
+                setNimbleTolerance(0)
+                setNimbleTestFolder("tests")
+                view = DynamicTypeView(frame: CGRect(x: 0, y: 0, width: 500, height: 500))
+            }
+
+            it("has a valid snapshot for all content size categories (iOS 9)") {
```
Comment (marcelofabri) [inline Bootstrap/BootstrapTests/DynamicTypeTests.swift]: I know I could use device agnostic screenshots here, but I didn't want to test that (yet)

File: Nimble-Snapshots.podspec (hunk 1 above)
Comment (ashfurrow) [inline Nimble-Snapshots.podspec]: I don't think so, but I'm fine removing it as a private header.

Comment (marcelofabri) [inline Nimble-Snapshots.podspec]: You mean leaving it public? If so, we probably want to prefix it just in case 😬

**Comment by marcelofabri** [comment]
On a related note: does it make sense to keep support Xcode 7.3? Can I remove it from Travis? 😬

**Comment by ashfurrow** [comment]
I'm even happy having this as an Xcode 8-only feature, whatever is easiest for you, really.

**Comment by marcelofabri** [comment]
I'm not sure what to do with Travis, it seems I'm getting a timeout when running the tests on simulator:

```
2016-09-29 20:43:38.595 xcodebuild[1741:5930] Error Domain=IDETestOperationsObserverErrorDomain Code=3 "Canceling tests due to timeout in Checking test manager availability... If you believe this error represents a bug, please attach the log file at /Users/travis/Library/Developer/Xcode/DerivedData/Bootstrap-dnfxxkeagupamkdwfuvsnzkcobxv/Logs/Test/52699553-06A4-4C99-A33D-DAAEFC7C3CFA/Session-BootstrapTests-2016-09-29_204131-YZOfct.log" UserInfo={NSLocalizedDescription=Canceling tests due to timeout in Checking test manager availability... If you believe this error represents a bug, please attach the log file at /Users/travis/Library/Developer/Xcode/DerivedData/Bootstrap-dnfxxkeagupamkdwfuvsnzkcobxv/Logs/Test/52699553-06A4-4C99-A33D-DAAEFC7C3CFA/Session-BootstrapTests-2016-09-29_204131-YZOfct.log}
```

**Comment by orta** [comment]
> On a related note: does it make sense to keep support Xcode 7.3? Can I remove it from Travis? 😬

If this project doesn't have a CHANGELOG, this is something we will need to put in it

**Comment by orta** [comment]
We do, so make sure this goes in it ;)

**Comment by orta** [comment]
Does this work?



Both of these images are set to 200px - text appears the same

File: Nimble-Snapshots.podspec [hunk 3]
```diff
@@ -12,9 +12,21 @@ Pod::Spec.new do |s|
...
+    ss.dependency "Quick"
+  end
+
+  s.subspec "DynamicType" do |ss|
+    ss.source_files  = "DynamicType/*.{swift,m,h}"
+    ss.frameworks = "UIKit"
+
+    ss.dependency "Nimble-Snapshots/Core"
+    ss.dependency "OCMock", "~> 3.3"
+  end
```
Comment (orta) [inline Nimble-Snapshots.podspec]: 👍

**Comment by marcelofabri** [comment]
@orta That actually is working - it seems that the Accessibility sizes only change Body fonts (got the clue from http://useyourloaf.com/blog/larger-dynamic-type-for-accessibility/). I'll change the tests to use that instead to be sure. Thanks for the catch!

**Comment by marcelofabri** [comment]
OK, so obviously that wouldn't work with Swift 3. I've made the required changes in aa0f9f9 and added another demo project using Swift 3 on 89024ce so we can be sure that it's working on both Swift versions.

I think this is ready for review as IMO the answers to my 2 questions in the original PR body is "not until someone needs it".

**Comment by orta** [comment]
Hrm, this fail seems completely arbitrary

```
2016-09-30 17:01:15.314 xcodebuild[1824:7468] Connection peer refused channel request for "dtxproxy:XCTestManager_IDEInterface:XCTestManager_DaemonConnectionInterface"; channel canceled <DTXChannel: 0x7fbabca734f0>
2016-09-30 17:01:15.338 xcodebuild[1824:5927] Error Domain=IDETestOperationsObserverErrorDomain Code=3 "The operation couldn’t be completed. (DTXProxyChannel error 1.) If you believe this error represents a bug, please attach the log file at /Users/travis/Library/Developer/Xcode/DerivedData/Bootstrap-dnfxxkeagupamkdwfuvsnzkcobxv/Logs/Test/6427A8BC-6487-4ECE-BA27-607004ACCF9D/Session-BootstrapTests-2016-09-30_170033-wXfVlp.log" UserInfo={NSLocalizedDescription=The operation couldn’t be completed. (DTXProxyChannel error 1.) If you believe this error represents a bug, please attach the log file at /Users/travis/Library/Developer/Xcode/DerivedData/Bootstrap-dnfxxkeagupamkdwfuvsnzkcobxv/Logs/Test/6427A8BC-6487-4ECE-BA27-607004ACCF9D/Session-BootstrapTests-2016-09-30_170033-wXfVlp.log}
▸ Touching BootstrapTests.xctest
▸ Signing /Users/travis/Library/Developer/Xcode/DerivedData/Bootstrap-dnfxxkeagupamkdwfuvsnzkcobxv/Build/Products/Debug-iphonesimulator/Bootstrap.app/PlugIns/BootstrapTests.xctest
** TEST FAILED **
```

**Comment by marcelofabri** [comment]
It seems that other people are having this issue too: https://github.com/travis-ci/travis-ci/issues/6422

**Comment by orta** [comment]
did you rebuild to make it green, or ship some changes?

**Comment by marcelofabri** [comment]
Not really, I thought you did it 😅

**Comment by orta** [comment]
Hah yeah, I do remember pressing it, never checked the results - I'm OK to just have spotty CI, PRs don't come along too often

**Comment by orta** [comment]
Then the only blocker for me is some changes to the README explaining the feature and that's good to go 👍

**Comment by marcelofabri** [comment]
Done in ed80ca8. Feel free to change the copy if you have any suggestions! 👍

**Comment by marcelofabri** [comment]
Is there anything else blocking this? 😅

**Comment by ashfurrow** [comment]
Cool! Thanks a lot, I've invited you to be a collaborator on the repo. I'll get a release of this out ASAP.

**Comment by ashfurrow** [comment]
Okay, pushed as 4.3.0 :tada:

**Comment by marcelofabri** [comment]
Awesome - thanks a lot Ash ❤️💯

**Pull request merged**